client/rpc/src/state/mod.rs

```diff
@@ -360,10 +379,98 @@ impl<Block, Client> State<Block, Client>
 			}.boxed()
 		})?;
 
+		module.register_subscription(
+			"state_runtimeVersion",
+			"state_unsubscribeRuntimeVersion",
+			|_params, mut sink, ctx| {
+				let executor = ctx.executor.clone();
+				let client = ctx.client.clone();
+
+				let mut previous_version = client.runtime_version_at(&BlockId::hash(client.info().best_hash))
+					.expect("best hash is valid; qed");
+				let _ = sink.send(&previous_version);
+				let rt_version_stream = client.storage_changes_notification_stream(Some(&[StorageKey(well_known_keys::CODE.to_vec())]), None, )
+					.map_err(|blockchain_err| Error::Client(Box::new(blockchain_err)))
+					.map_err(to_jsonrpsee_call_error)?;
+
+				let fut = async move {
+					let mut stream = rt_version_stream
+						.filter_map(move |_| {
+							let info = client.info();
+							let version = client
+								.runtime_version_at(&BlockId::hash(info.best_hash));
+								match version {
+									Ok(v) => if previous_version != v {
+											previous_version = v.clone();
+											future::ready(Some(v))
+										} else {
+											future::ready(None)
+										},
+									Err(e) => {
+										log::error!("Could not fetch current runtime version. Error={:?}", e);
+										// TODO: Does this terminate the stream? What is the best way to let users know?
+										future::ready(None)
+									}
+								}
+						});
+						stream.for_each(|version| {
+							match sink.send(&version) {
+								Ok(_) =>  future::ready(()),
+								Err(e) => {
+									log::error!("Could not send data to the state_runtimeVersion subscriber: {:?}", e);
+									future::ready(())
+								},
+							}
+						}).await;
+				}.boxed();
+				executor.execute_new(fut);
+				Ok(())
+		})?;
 
-		// TODO: add subscriptions.
+		module.register_subscription(
+			"state_storage",
+			"state_unsubscribeStorage",
+			|params, mut sink, ctx| {
+				let executor = ctx.executor.clone();
+				let keys = params.one::<Option<Vec<StorageKey>>>()?;
+				let stream = ctx.client.storage_changes_notification_stream(
+					keys.as_ref().map(|keys| &**keys),
+					None
+				).unwrap();
+				let fut = async move {
+					let stream = stream.map(|(block, changes)| {
+						StorageChangeSet {
+							block,
+							changes: changes
+								.iter()
+								.filter_map(|(o_sk, k, v)| {
+								// Note: the first `Option<&StorageKey>` seems to be the parent key, so it's set only
+								// for storage events stemming from child storage, `None` otherwise. This RPC only
+								// returns non-child storage.
+								if o_sk.is_none() {
+									Some((k.clone(), v.cloned()))
+								} else {
+									None
+								}
+							}).collect(),
+						}
+					});
+					stream.for_each(|data| {
```

Comment:
I assume this will run forever, maybe `stream.take_while(|data| sink.send(&data).is_ok())` to close it when the client is not interested anymore?

Reply:
I tried that and got ```error[E0277]: `bool` is not a future``` which I'm not sure how to fix. I'll return to it tomorrow.

Reply:
What I would really like is `stream.take_while(|data| sink.send(&data))` and have the error propagated back and the stream closed.

Reply:
ah, my bad it must be a future that returns bool

Reply:
About `.for_each(|_| future::ready(()))` – do you think `.into_future()` works there? (It compiles...)

Reply:
no, it will only fetch one element from stream then, I know it looks a bit noisy